library/kotlin/src/io/envoyproxy/envoymobile/GRPCResponseHandler.kt

```diff
@@ -0,0 +1,196 @@
+   *
+   * @param closure: Closure which will be called when the stream is canceled.
+   * @return ResponseHandler, this ResponseHandler.
+   */
+  fun onCanceled(closure: () -> Unit): GRPCResponseHandler {
+    underlyingHandler.onCanceled(closure)
+    return this
+  }
+
+  /**
+   * Recursively processes a buffer of data, buffering it into messages based on state.
+   * When a message has been fully buffered, `onMessage` will be called with the message.
+   *
+   * @param bufferedStream The buffer of data from which to determine state and messages.
+   * @param processState The current process state of the buffering.
+   * @param onMessage Closure to call when a new message is available.
+   */
+  private fun processData(
+      bufferedStream: ByteArrayOutputStream,
+      processState: ProcessState,
+      onMessage: (byteBuffer: ByteBuffer) -> Unit): ProcessState {
+
+    var nextState = processState
+
+    when (processState) {
+      is ProcessState.CompressionFlag -> {
+        val byteArray = bufferedStream.toByteArray()
+        if (byteArray.size < MESSAGE_HEADING_OFFSET) {
+          // We don't have enough information so we'll just return
+          return ProcessState.CompressionFlag
+        }
+
+        val compressionFlag = byteArray[0]
+        // TODO: Support gRPC compression https://github.com/lyft/envoy-mobile/issues/501
+        if (compressionFlag.compareTo(0) != 0) {
+          errorClosure(
+              EnvoyError(
+                  EnvoyErrorCode.ENVOY_UNDEFINED_ERROR,
+                  "Unable to accept compression enabled messages for gRPC"))
+
+          // no op the current onData and clean up
+          underlyingHandler.onData { _, _ -> }
+          bufferedStream.reset()
+        }
+
+        nextState = ProcessState.MessageLength
+      }
+      is ProcessState.MessageLength -> {
+        val byteArray = bufferedStream.toByteArray()
+        val messageLength = ByteBuffer.wrap(byteArray.sliceArray(1..4)).int
+
+        if (messageLength == 0) {
+          onMessage(ByteBuffer.wrap(ByteArray(0)))
+          return ProcessState.CompressionFlag
+        }
+
+        nextState = ProcessState.Message(messageLength)
+      }
+      is ProcessState.Message -> {
+        if (bufferedStream.size() < processState.messageLength + MESSAGE_HEADING_OFFSET) {
+          // We don't have enough bytes to construct the message so we'll return
+          return ProcessState.Message(processState.messageLength)
+        }
+
+        val byteArray = bufferedStream.toByteArray()
+        onMessage(ByteBuffer.wrap(
+            byteArray.sliceArray(MESSAGE_HEADING_OFFSET until MESSAGE_HEADING_OFFSET + processState.messageLength)))
+        bufferedStream.reset()
+        bufferedStream.write(
+            byteArray.sliceArray(MESSAGE_HEADING_OFFSET + processState.messageLength until byteArray.size))
+
+        if (byteArray.sliceArray(MESSAGE_HEADING_OFFSET + processState.messageLength until byteArray.size).isEmpty()) {
```

Comment:
I'm actually more inclined to have early exits and returns but the Swift implementation has the `processData` at the very end so I tried to keep this style closer to that. 

IMO, I do think it'll be clearer to return on each switch case since we wont be hitting multiple cases ever. If you agree with this, we can make changes to the Swift code to do that and I'll make the changes here too